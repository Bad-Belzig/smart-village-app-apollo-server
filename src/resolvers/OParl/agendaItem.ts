import {
  AgendaItem,
  Consultation,
  File,
  IAgendaItem,
  Meeting,
} from '../../models';

export const agendaItemResolvers = {
  Query: {
    oParlAgendaItems: () => AgendaItem.find(),
  },
  OParlAgendaItem: {
    meeting: (args: IAgendaItem) =>
      Meeting.findOne({ externalId: args.meeting }),
    consultation: (args: IAgendaItem) =>
      Consultation.findOne({ externalId: args.consultation }),
    resolutionFile: (args: IAgendaItem) =>
      File.findOne({ externalId: args.resolutionFile }),
    auxiliaryFile: (args: IAgendaItem) =>
      args.auxiliaryFile?.map((value) => File.findOne({ externalId: value })),
  },
};
