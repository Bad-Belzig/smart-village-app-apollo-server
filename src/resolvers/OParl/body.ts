import {
  Body,
  IBody,
  LegislativeTerm,
  Location,
  Meeting,
  Organization,
  Paper,
  Person,
} from '../../models';

export const bodyResolvers = {
  Query: {
    oParlBodies: () => Body.find(),
  },
  OParlBody: {
    organization: (args: IBody) =>
      args.organization.map((value) =>
        Organization.findOne({ externalId: value }),
      ),
    person: (args: IBody) =>
      args.person.map((value) => Person.findOne({ externalId: value })),
    meeting: (args: IBody) =>
      args.meeting.map((value) => Meeting.findOne({ externalId: value })),
    paper: (args: IBody) =>
      args.paper.map((value) => Paper.findOne({ externalId: value })),
    legislativeTerm: (args: IBody) =>
      args.legislativeTerm.map((value) =>
        LegislativeTerm.findOne({ externalId: value }),
      ),
    location: (args: IBody) => Location.findOne({ externalId: args.location }),
  },
};
