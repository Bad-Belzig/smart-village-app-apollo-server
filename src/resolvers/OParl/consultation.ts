import { FilterQuery } from 'mongoose';

import {
  AgendaItem,
  Consultation,
  IConsultation,
  IConsultationSchema,
  Meeting,
  Organization,
  Paper,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const consultationResolvers = {
  Query: {
    oParlConsultations: (
      _,
      args: { externalIds?: string[]; keyword?: string[] },
    ) => {
      const filter: FilterQuery<IConsultationSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return Consultation.find(filter);
    },
  },
  OParlConsultation: {
    paper: (parent: IConsultation) =>
      Paper.findOne({ externalId: parent.paper }),
    agendaItem: (parent: IConsultation) =>
      AgendaItem.findOne({ externalId: parent.agendaItem }),
    meeting: (parent: IConsultation) =>
      Meeting.findOne({ externalId: parent.meeting }),
    organization: async (
      parent: IConsultation,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Organization,
        parent.organization,
        args.offset,
        args.pageSize,
      ),
  },
};
