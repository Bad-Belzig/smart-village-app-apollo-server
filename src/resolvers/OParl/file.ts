import { FilterQuery } from 'mongoose';

import {
  AgendaItem,
  File,
  IFile,
  IFileSchema,
  Meeting,
  Paper,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const fileResolvers = {
  Query: {
    oParlFiles: (_, args: { externalIds?: string[]; keyword?: string[] }) => {
      const filter: FilterQuery<IFileSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return File.find(filter);
    },
  },
  OParlFile: {
    masterFile: (parent: IFile) =>
      File.findOne({ externalId: parent.masterFile }),
    derivativeFile: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        File,
        parent.derivativeFile,
        args.offset,
        args.pageSize,
      ),
    meeting: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Meeting,
        parent.meeting,
        args.offset,
        args.pageSize,
      ),
    agendaItem: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        AgendaItem,
        parent.agendaItem,
        args.offset,
        args.pageSize,
      ),
    paper: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(Paper, parent.paper, args.offset, args.pageSize),
  },
};
