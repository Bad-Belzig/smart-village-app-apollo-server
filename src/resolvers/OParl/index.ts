export * from './agendaItem';
export * from './body';
export * from './consultation';
export * from './file';
export * from './keywords';
export * from './legislativeTerm';
export * from './location';
export * from './meeting';
export * from './membership';
export * from './organization';
export * from './paper';
export * from './person';
export * from './system';
