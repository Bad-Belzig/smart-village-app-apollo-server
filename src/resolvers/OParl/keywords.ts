import { isStringArray } from '../../helpers';
import { KeywordKey, Keywords } from '../../models';

export const keywordsResolver = {
  Query: {
    oParlKeywordList: async () => {
      const keywords = await Keywords.findOne().exec();

      if (!keywords) {
        return [];
      }

      const res = new Set<string>();

      for (const key in KeywordKey) {
        const keywordsForKey = keywords[key];
        if (isStringArray(keywordsForKey)) {
          keywordsForKey.forEach((keyword) => res.add(keyword));
        }
      }

      return [...res].sort((a, b) =>
        a.toLowerCase() <= b.toLowerCase() ? -1 : 1,
      );
    },
  },
};
