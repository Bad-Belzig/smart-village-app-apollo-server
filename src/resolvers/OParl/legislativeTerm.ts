import { Body, ILegislativeTerm, LegislativeTerm } from '../../models';
import { findByIds } from '../resolverHelpers';

export const legislativeTermResolvers = {
  Query: {
    oParlLegislativeTerms: (_, args: { externalIds?: string[] }) =>
      args.externalIds
        ? findByIds(args.externalIds, LegislativeTerm)
        : LegislativeTerm.find(),
  },
  OParlLegislativeTerm: {
    body: (parent: ILegislativeTerm) =>
      Body.findOne({ externalId: parent.body }),
  },
};
