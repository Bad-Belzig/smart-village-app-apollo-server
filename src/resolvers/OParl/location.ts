import {
  Body,
  ILocation,
  Location,
  Meeting,
  Organization,
  Paper,
} from '../../models';

export const locationResolvers = {
  Query: {
    oParlLocations: () => Location.find(),
  },
  OParlLocation: {
    bodies: (args: ILocation) =>
      args.bodies?.map((value) => Body.findOne({ externalId: value })),
    organization: (args: ILocation) =>
      args.organization?.map((value) =>
        Organization.findOne({ externalId: value }),
      ),
    meeting: (args: ILocation) =>
      args.meeting?.map((value) => Meeting.findOne({ externalId: value })),
    papers: (args: ILocation) =>
      args.papers?.map((value) => Paper.findOne({ externalId: value })),
  },
};
