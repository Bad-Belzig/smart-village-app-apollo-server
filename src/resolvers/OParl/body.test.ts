import { MongoMemoryServer } from 'mongodb-memory-server';
import mongoose from 'mongoose';
import { importBody } from '../../services/OParl/body';

const mongod = new MongoMemoryServer();

const databaseName = 'test_body_resolver';

// TODO: change test url to something within our control
const testUrl = 'https://www.muenchen-transparent.de/oparl/v1.0/body/0';

describe('importing a body', () => {
  beforeAll(async () => {
    const uri = await mongod.getUri(databaseName);

    await mongoose.connect(uri, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
      keepAlive: true,
    });
  });

  it('imports properly', async () => {
    const body = await importBody(testUrl);
    expect(body).toBeDefined();
    const json = body?.toJSON();
    delete json?._id;
    expect(json).toMatchSnapshot();
  });

  it('import returns undefined on faulty url', async () => {
    const body = await importBody(
      'https://smart-village.solutions/smart-village-solutions/',
    );
    expect(body).toBeUndefined();
  });

  afterAll(async () => {
    await mongoose.connection.dropDatabase();
    await mongoose.connection.close();
    mongod.stop();
  });
});
