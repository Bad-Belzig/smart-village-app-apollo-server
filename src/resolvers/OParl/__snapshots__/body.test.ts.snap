// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a body imports properly 1`] = `
Object {
  "__v": 0,
  "ags": "09162000",
  "contactEmail": "info@muenchen-transparent.de",
  "contactName": "München Transparent",
  "created": 2016-10-16T18:38:18.000Z,
  "equivalent": Array [],
  "externalId": "https://www.muenchen-transparent.de/oparl/v1.0/body/0",
  "keyword": Array [],
  "legislativeTerm": Array [
    "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/0",
    "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/1",
    "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/2",
    "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/3",
    "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/4",
  ],
  "meeting": "https://www.muenchen-transparent.de/oparl/v1.0/body/0/list/meeting",
  "modified": 2016-10-16T18:39:16.000Z,
  "name": "Stadrat der Landeshauptstadt München",
  "organization": "https://www.muenchen-transparent.de/oparl/v1.0/body/0/list/organization",
  "paper": "https://www.muenchen-transparent.de/oparl/v1.0/body/0/list/paper",
  "person": "https://www.muenchen-transparent.de/oparl/v1.0/body/0/list/person",
  "rgs": "606109162000",
  "type": "https://schema.oparl.org/1.0/Body",
  "web": "https://www.muenchen-transparent.de/",
  "website": "https://www.muenchen.de/rathaus/",
}
`;
